### ISSUE TITLE
[PyTorch] indexing by a zero-dim Tensor should slice (not copy)

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#6217]
PyTorch master

Currently, indexing a tensor by a zero-dim tensor creates a copy. We should make it return an alias.

```python
idx = torch.tensor(0, dtype=torch.int64)
x = torch.arange(9).reshape(3, 3)
x[idx].zero_()   # no effect since x[idx] copies; it should slice instead
print(x)
```

```
 0  1  2
 3  4  5
 6  7  8
[torch.FloatTensor of size (3,3)]
```

Reported by lvdmaaten

**Comment by gchanan** [comment]
@colesbury:
Should this block 0.4?

Also, do we have a justification for this behavior we could add to the docs?  It seems what you are suggesting here is less unexpected, because people think of 0-dimensional tensors like python numbers, but one could make the case the other way too (and we did when we first implemented this).

**Comment by colesbury** [comment]
Yeah, I think it should block 0.4.

It seems like that when the behavior of zero-dim arrays and NumPy scalars differ, we should prefer behaving like scalars, since people have more experience with them.

**Comment by gchanan** [comment]
The only annoying thing about that is right now it's easy in general to get the scalar behavior -- just call `.item()`.  Now I need to think about how to get the tensor behavior -- do I unsqueeze?  do I copy? do I clone?  Presumably the right thing to do is dependent on the actual operation.

**Comment by apaszke** [comment]
I think it's better to use the sharing version of the op as often as we can. This is what one usually expects, and is also needed for backwards compatibility.
